Feature/parselmouth fix

JoFrhwld opened the pull request.

JoFrhwld committed:
cherry pickk parselmouth initial working commit

JoFrhwld committed:
cherry-pick pulled praat call from aligner.py

JoFrhwld committed:
final parsemout refactor. closes #68

chrisbrickhouse committed:
Merge dev into feature/parselmouth-fix

chrisbrickhouse reviewed (commented):
So my top-line point is that I would object to merging this for a while, and I think we should move it from a 2.1 milestone to 2.2.

The big issue is around deprecation. This is quite a big change to the code-base, and might have unforseen effects on workflows. For example, we technically support ESPS as an alternative to praat. I don't know who uses that (if anyone), but this patch removes that support silently. That's not good for stability. Additionally, if we want to convince the DARLA team to update more frequently (and report upstream) we should make some guarantees to downstream in terms of notification and stability. So with that in mind, I think we should use this as an opportunity to develop and implement a deprecation policy.

I recommend reading [MediaWiki's deprecation policy](https://www.mediawiki.org/wiki/Stable_interface_policy#Deprecation_process). At the very least, for this patch, I recommend we provide deprecation warnings for these in 2.1, and spend the 2.2 dev cycle hammering this out further.

Which gets to my second point: code quality. For example, a lot of `praat.py` extends parselmouth classes. We should be using subclasses instead of the interface developed here, but that will take a fair bit more work. We should also write a lot more tests. Migrating to parselmouth could effect results and analysis for users, so we want to make *very* sure that the core features are equivalent. We should have some feature tests to show that the results don't change. That's a lot to write, so delaying until 2.2 gives us that time.

Despite all the above, I think this is a very good patch, and don't want it to seem like I don't appreciate it. It's just that this has turned out to be quite a big change, and we should hammer out a roadmap to make sure we avoid unintended problems.

JoFrhwld commented:
All good points. Re: speech software deprecation, I'm not sure we need to be worried about dropping esps support since, as far as I can tell, it hasn't been actually acquirable software for about a decade or more (all links to it 404).

But, I think retaining `checkSpeechSoftware`, perhaps using parselmouth as a fallback or its own specifiable option is a good idea. Without wholesale deprecating the praat option & adding parselmouth as an option, this could probably still make it into 2.1? All that writing to disc is a real blocker to any kind of parallelization, which is why I'm eager to firmly set us down this road.

JoFrhwld commented on fave/extractFormants.py:
```diff
@@ -62,6 +62,7 @@
 """
 
 
+from curses.ascii import SO
```
This is bad, but, I'm not 100% sure why this is here.

chrisbrickhouse replied:
Hmm seems to be a library for working with ascii, and we're importing the "shift out" character which denotes the start of an alternate character set ([docs](https://docs.python.org/3/library/curses.ascii.html)). It doesn't seem like we're using it, so if you don't know what it's for, we may as well remove it and save ourselves from figuring out the test failures.

chrisbrickhouse committed:
Add deprecation for SoX and Praat methods.

See #85 and #68

JoFrhwld commented on fave/praat.py:
```diff
@@ -1,40 +1,57 @@
+import parselmouth
 
 class Formant:
 
     """represents a formant contour as a series of frames"""
 
-    def __init__(self, name=None):
+    def __init__(self, name=None, formant = None, maxFormant = None):
+        if formant and maxFormant:
+            if isinstance(formant, parselmouth.Formant):
+                self.__times = list(formant.ts())  # list of measurement times (frames)
+                self.__intensities = []
+                    # list of intensities (maximum intensity in each frame)
+                self.__formants = [[formant.get_value_at_time(formant_number = N, time = T) 
+                                      for N in (1,2,3) ]  
+                                    for T in formant.ts()]
+                    # list of formants frequencies (F1-F3, for each frame)
+                self.__bandwidths = [[formant.get_bandwidth_at_time(formant_number = N, time = T) 
+                                        for N in (1,2,3)] 
+                                    for T in formant.ts()]
+                    # list of bandwidths (for each formant F1-F3, for each frame)
+                                            # !!! CHANGED:  all above lists no longer include frames with only
+                                            # a minimum of 2 formant measurements
+                                            # !!!
+                self.__xmin = formant.xmin  # start time (in seconds)
+                self.__xmax = formant.xmax  # end time (in seconds)
+                self.__nx = formant.nx  # number of frames
+                self.__dx = formant.dx  # time step = frame duration (in seconds)
+                self.__x1 = formant.x1  # start time of first frame (in seconds)
+                self.__maxFormants = maxFormant  # maximum number of formants in a frame
```
This could be rolled out into a separate module, like the `praat.Formant` and `esps.Formant` modules are. I'd have to really think about how to make it a real subclass of parselmouth.Formant, since the formant class is [written in C++](https://github.com/YannickJadoul/Parselmouth/blob/master/src/parselmouth/Formant.cpp).

chrisbrickhouse replied:
Yeah, I'm realizing that subclassing isn't going to work well given how the formant class comes to us---we don't create it, rather, it's returned from a function call. I started building a bridge in the most recent commits. See af8f871f8ec83f5bfe48128604e75ad050c3face

chrisbrickhouse commented:
> All good points. Re: speech software deprecation, I'm not sure we need to be worried about dropping esps support since, as far as I can tell, it hasn't been actually acquirable software for about a decade or more (all links to it 404).

Eh, I guess I can get behind removal. HTK's also been moribund for a while which is where my hesitation comes from, but that's at least kinda available---ESPS isn't. As long as we're super clear about the change in the release notes, I think you're right that this is low risk. 

> But, I think retaining `checkSpeechSoftware`, perhaps using parselmouth as a fallback or its own specifiable option is a good idea. Without wholesale deprecating the praat option & adding parselmouth as an option, this could probably still make it into 2.1?

I've added some deprecation warnings in 83cbde6 to that effect. Could be specified more (and probably include a `@versionchanged` decorator).

> All that writing to disc is a real blocker to any kind of parallelization, which is why I'm eager to firmly set us down this road.

Hmm, that feature wasn't clear to me from the PR. I agree that's a priority and should be included in 2.1. I think the combo of deprecation, esps removal, and parselmouth inclusion is a good one. The subclassing and such can be documented as technical debt and fixed later.

JoFrhwld commented:
I haven't really messed around with the Github Project kanban boards since I set them up, but could be useful for planning a little bit of this out. https://github.com/users/JoFrhwld/projects/1/views/1

What I've managed to figure out about the history of why ESPS was included at all is from here: http://www.phon.ox.ac.uk/releases Specifically

> Subsequently, Stephen Isard (Institute for Research in Cognitive Science, University of Pennsylvania) updated the sources to compile and run on current OSX and Linux machines, and Danny Yee (University of Oxford) has packaged them as [a .deb file for Ubuntu 12.04](http://www.phon.ox.ac.uk/files/releases/esps_60.6.4-0_all.deb). The licence terms are exactly those of the final ESPS source code release, without modification.

Stephan Isard was a co-author with Keelan Evanini and Mark Liberman on the first FAVE-like system in 2008, and Ubuntu 12.04 was initially released in 2012, all immediately prior to the first verisons of the FAVE-suite being created.

chrisbrickhouse committed:
Restore praat scripts (#89)

chrisbrickhouse committed:
Revert "Merge dev into feature/parselmouth-fix"

This reverts commit a429f1595866221302ba082c00a5f87c09b32f92, reversing
changes made to 7bf79e94b13c0139c313a5214cf24c9a65d63592.

chrisbrickhouse committed:
Merge dev into feature/parselmouth-fix

chrisbrickhouse committed:
Move parselmouth logic to new parselmouth_bridge

Creates the new parselmouth_bridge module and deprecates
some of the fave.praat module in favor of these parselmouth-based
classes.

chrisbrickhouse commented on fave/align/transcriptprocessor.py:
```diff
@@ -247,13 +247,7 @@ def read_transcription_file(self):
-                # Log a warning about having detected a header row
-                self.logger.warning('Header row was detected')
-                del lines[0]
+        self.lines = lines
```
This reverts the header detection from #65, we should remember to fix this before merging

chrisbrickhouse committed:
Restore removed test files

JoFrhwld committed:
moved over to pm bridge

JoFrhwld committed:
pm cross over debug

JoFrhwld committed:
Gotta do numpy checks rather than length

JoFrhwld committed:
trying to squash these bugs

JoFrhwld committed:
More np.nan bug fixes

keelanevanini commented:
@JoFrhwld I realize that I'm quite late to this discussion, but I just stumbled upon it and I thought I would chime in with my perspective on the inclusion of ESPS in the FAVE toolkit. You are correct that I added ESPS to the early version of the `extractFormants.py` script around 2007-2008 based on a recommendation from either Stephen Isard and Mark Liberman (I can't remember which one -- the three of us had regular meetings to discuss approaches to automated formant measurement). I generally preferred using ESPS for extracting formants over Praat, since the ESPS `format` command could be called easily from shell scripts and I found Praat scripts to be cumbersome. I used ESPS in the `extractFormants.py` script as the early precursor to FAVE in my dissertation research and publications that stemmed from it (such as this Interspeech paper: https://www.isca-archive.org/interspeech_2009/evanini09_interspeech.pdf). 

However, even back then, it was not straightforward to obtain the ESPS source code and compile it. I'm guessing that I was probably the only FAVE user ever to use ESPS instead of Praat for formant extraction. So, from that perspective, I think that deprecating ESPS in FAVE is no-risk. I approve :-)
